### PR TITLE
Stackanetes uses kpm & Ingress.

### PR DESCRIPTION
**Pull request by Quentin-M**
**ChangeLog** (big lines)
- Based of 3c00aa8858652e
- Discard the old python deployer and replace it by [kpm](https://github.com/coreos/kpm)
  - Provide many advantages in term of lifecycle management (updates) and flexibility (conditional deployments)
  - Ease code maintenance, modularity and reusability with a simple architecture made of pretty independent packages and human-readable Kubernetes manifests 
  - Add ability to easility deploy sharded applications
  - Reduce the list of dependencies for deployment to one
  - Allow reproducible deployments with versioning, support for release channels and sharable configuration 
- Add support for external load-balancer to access OpenStack APIs (formerly https://github.com/stackanetes/stackanetes/pull/73)
- Remove rados-gateway as it should be deployed as part of Ceph instead (as discussed) and make Glance use rbd instead, therefore adding Copy-on-Write support
- Remove fluentd/kibana/elastisearch as it should be deployed as part as Kubernetes instead (as discussed)
- Make searchlight use a standard elastic search image as the one provided by Kubernetes has too several hardcoded values and poor [cluster bootstrapping behavior](https://github.com/kubernetes/kubernetes/blob/ef0c9f0/cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go#L61) (as discussed)
- Remove usage of the Ceph's admin keyring and use specific users instead
- Offer ability to configure multiple DNS servers, which may be useful to resolve Kubernetes nodes' hostnames to enable live-migrations
- Rewrite README entirely with clear requirements, setup instructions and expectations
- Provide OpenStack bash files for testing/demo purposes
- Simplify and cleanup where clearly feasible,
- Adopt some standards around manifest definition

### CODE REVIEW

File: nova/templates/configmaps/nova.conf.yaml [hunk 1]
```diff
@@ -0,0 +1,129 @@
+apiVersion: v1
+kind: ConfigMap
+data:
+  nova.conf: |+
+    [DEFAULT]
+    debug = {{ misc.debug }}
+    vif_plugging_is_fatal = False
```
Comment (ss7pro) [inline nova/templates/configmaps/nova.conf.yaml]: Please recover '#FIXME' sections so we will know what needs to be fixed.

**Comment by PiotrProkop** [comment]
Can we have a way to somehow template passwords ? For now it is just hard-coded.

File: nova/templates/configmaps/nova.conf.yaml (hunk 1 above)
Comment (Quentin-M) [inline nova/templates/configmaps/nova.conf.yaml]: Thank you. I fixed that one and verified manually that all other FIXME were there.

**Comment by Quentin-M** [comment]
@PiotrProkop Values (incl. passwords) in `stackanetes/parameters.yaml` aren't hardcoded per se. They are the ones that are shipped by default within the Stackanetes package to the kpm registry and thus, they serve as defaults. Any values can be overridden at deploy time using a YAML file containing custom values, using ```kpm deploy stackanetes/stackanetes --namespace openstack --variables custom.yaml -H https://beta.kpm.sh``` (as in the README).

```
# custom.yaml
keystone:
  admin_password: foo
```

Does that make sense?

File: README.md [hunk 2]
```diff
@@ -1,144 +1,176 @@
...
+### Requirements
+
+Stackanetes requires Kubernetes 1.3+ with:
+  - At least two schedulable nodes,
+  - At least one virtualization-ready node,
+  - [Overlay network] & [DNS add-on],
+  - Kubelet running with `--allow-privileged=true`,
+
+While Glance may operate with local storage, a Ceph cluster is needed for Cinder.
+Nova's live-migration feature also requires Ceph and DNS resolution of the Kubernetes nodes' hostnames.
```
Comment (ss7pro) [inline README.md]: CEPH is not necessary for live migration, but it's required for full HA (ie: for glance).

File: cinder/manifest.jsonnet [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+        named_host: $.variables.network.ingress.host % "volume",
+      },
+    },
+
+    database: {
+      address: "mariadb",
+      port: 3306,
+      root_user: "root",
+      root_password: "password",
```
Comment (ss7pro) [inline cinder/manifest.jsonnet]: Hardcoded database password, it should be set in a configuration file.

File: cinder/manifest.jsonnet [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        port: 30080,
+
+        named_host: $.variables.network.ingress.host % "volume",
+      },
+    },
+
+    database: {
+      address: "mariadb",
+      port: 3306,
+      root_user: "root",
```
Comment (ss7pro) [inline cinder/manifest.jsonnet]: hardcoded database user, this should be set in a configuration file.

File: cinder/manifest.jsonnet [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    expander: "jinja2",
+    author: "Quentin Machu",
+    version: "0.1.0",
+    description: "cinder",
+    license: "Apache 2.0",
+  },
+
+  variables: {
+    deployment: {
+      node_label: "openstack-control-plane",
```
Comment (ss7pro) [inline cinder/manifest.jsonnet]: This label name should be moved to a central configutration file. So we don't need to change it in each component separetley.

File: cinder/manifest.jsonnet [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+      replicas: 1,
+
+      image: {
+        base: "quay.io/stackanetes/stackanetes-%s:barcelona",
```
Comment (ss7pro) [inline cinder/manifest.jsonnet]: same as above, this should be keept in a single central configuration file.

File: cinder/manifest.jsonnet [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+      },
+    },
+
+    network: {
+      ip_address: "{{ .IP }}",
+      port: 8776,
+
+      ingress: {
+        enabled: true,
+        host: "%s.openstack.cluster",
```
Comment (ss7pro) [inline cinder/manifest.jsonnet]: domain name should be also taken from a central configutration file.

File: cinder/manifest.jsonnet (hunk 3 above)
Comment (ant31) [inline cinder/manifest.jsonnet]: `manifest.jsonnet` is the configuration file, the values here are the default that can be changed via "stackanetes/parameters.yaml' or via the command line.

Comment (Quentin-M) [inline cinder/manifest.jsonnet]: It is configurable in `stackanetes/parameters.yaml` (or more generally in any YAML file specified at deploy time with `--variables custom.yaml`.

File: cinder/manifest.jsonnet (hunk 4 above)
Comment (Quentin-M) [inline cinder/manifest.jsonnet]: ditto

File: cinder/manifest.jsonnet (hunk 5 above)
Comment (Quentin-M) [inline cinder/manifest.jsonnet]: ditto. I'll add it to the default `parameters.yaml` so everyone can see it and edit it without searching - but even if it isn't there, any variables can be overridden by specifying a YAML file at deploy time.

File: cinder/manifest.jsonnet (hunk 7 above)
Comment (Quentin-M) [inline cinder/manifest.jsonnet]: ditto :)

File: README.md (hunk 2 above)
Comment (Quentin-M) [inline README.md]: Thanks, updated :)

File: cinder/manifest.jsonnet [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    },
+
+    {
+      file: "db-sync.yaml",
+      template: (importstr "templates/db-sync.yaml"),
+      name: "cinder-db-sync",
+      type: "job",
+    },
+
+    // Daemons.
```
Comment (DTadrzak) [inline cinder/manifest.jsonnet]: s/Daemons/Deployments/

File: glance/manifest.jsonnet [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    },
+
+    {
+      file: "post.yaml",
+      template: (importstr "templates/post.yaml"),
+      name: "glance-post",
+      type: "job",
+    },
+
+    // Daemons.
```
Comment (DTadrzak) [inline glance/manifest.jsonnet]: s/deployment/daemon/

File: mariadb/manifest.jsonnet [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    },
+
+    {
+      file: "configmaps/start.sh.yaml",
+      template: (importstr "templates/configmaps/start.sh.yaml"),
+      name: "mariadb-startsh",
+      type: "configmap",
+    },
+
+    // Daemons.
```
Comment (DTadrzak) [inline mariadb/manifest.jsonnet]: Deployments

**Comment by PiotrProkop** [comment]
LGTM

**Pull request merged**